modules/core/src/main/scala/org/tessellation/infrastructure/snapshot/SnapshotService.scala

```diff
@@ -27,48 +29,57 @@ object SnapshotService {
 
   def make[F[_]: Async](
     snapshotStorage: SnapshotStorage[F],
-    blocksQueue: Queue[F, Set[Signed[DAGBlock]]],
+    snapshotPreconditionsValidator: SnapshotPreconditionsValidator[F],
+    l1DataQueue: Queue[F, L1Data],
     config: SnapshotConfig
   ): SnapshotService[F] =
     new SnapshotService[F] {
 
-      val snapshotCreation: Stream[F, GlobalSnapshot] =
+      val pipeline: Stream[F, Either[Signed[DAGBlock], SnapshotReference]] =
         Stream
-          .fromQueueUnterminated(blocksQueue)
-          .evalMap(createSnapshot)
-          .through(switchRepeat(config.fallbackTriggerTimeout, Stream.eval(createSnapshot())))
+          .fromQueueUnterminated(l1DataQueue)
+          .zip(Stream.eval(snapshotStorage.getLastSnapshotHeight))
+          .evalMap {
+            case (data, height) =>
+              snapshotPreconditionsValidator.validate(height, data) match {
+                case Valid(height) =>
+                  createSnapshotTrigger(height).map(_.asRight[Signed[DAGBlock]])
```

Comment:
shouldn't it send both: `data.block` and then `snapshotTrigger`?

Reply:
Fair point!